packages/teleterm/src/ui/TabHost/useDocumentCreator.ts

```diff
@@ -0,0 +1,59 @@
+import { MutableRefObject, useCallback } from 'react';
+import { Document } from '../types';
+import uris from '../uris';
+import { unique } from '../utils/uid';
+import DocumentService from '../services/docs';
+
+export interface DocumentTerminalRef {
+  type: 'document_terminal';
+
+  getWorkingDirectory(): Promise<string>;
+}
+
+export type ActiveDocumentRef = DocumentTerminalRef;
+
+export function useDocumentCreator(options: {
+  serviceDocs: DocumentService;
+  activeDocumentRef: MutableRefObject<ActiveDocumentRef>;
+}) {
+  async function getTerminalWorkingDirectory(): Promise<string | undefined> {
+    if (options.activeDocumentRef.current.type === 'document_terminal') {
+      try {
+        return await options.activeDocumentRef.current.getWorkingDirectory();
+      } catch (err) {
+        return; // default cwd will be used
+      }
+    }
+  }
+
+  async function getNewDocumentBasingOnActive(): Promise<Document> {
```

Comment:
Lets do this. Instead of passing around the document to grab an instance of pty to execute a method that returns a directory, instead we can do this (sample below).

```tsx
export default function createPtyService(runtimeSettings: RuntimeSettings) {
  
  const service = {

    getProcessDirectory(pid){
      return PtyProcess.getDirectory(pid)
    },

    // the rest
```

```tsx
export default function useDocumentTerminal(doc: Props['doc']) {
  const ctx = useAppContext();

  const ptyProcess = React.useMemo(() => {
    const activeDoc = ctx.serviceDocs.getActive()        
    if (activeDoc.kind === 'terminal_tsh_session' || 'terminal_shell'){
      cwd = ctx.serviceTerminals.getDirectory(activeDoc.pid);
    }

    if (doc.kind === 'terminal_tsh_session') {        
      return ctx.serviceTerminals.createPtyProcess({
        ...doc,
        cwd,
        kind: 'tsh-login',
      });
    }

    if (doc.kind === 'terminal_shell') {      
      return ctx.serviceTerminals.createPtyProcess({ kind: 'new-shell', cwd, });
    }
  }, []);

  React.useEffect(() => {
    ctx.serviceDocs.update(doc.uri, { pid:  ptyProcess.getPid() })

    const cleanup = () => {
      ptyProcess.dispose();
    };

    return cleanup;
  }, [ptyProcess]);

  return {
    ptyProcess,
  };
}
```

Reply:
Btw, we need to think how we can do the same for `tsh` session (getting to know the platform host is not trivial)